python/tank_vendor/shotgun_deploy/io_descriptor/factory.py

```diff
@@ -36,12 +46,34 @@ def create_io_descriptor(sg, descriptor_type, dict_or_uri, bundle_cache_root, fa
     from .git_branch import IODescriptorGitBranch
     from .manual import IODescriptorManual
 
+
+    # first check if we already have this in our cache
+    # Since all our normal descriptors are immutable - they represent a specific,
+    # read only and cached version of an app, engine or framework on disk, we can
+    # also cache their wrapper objects.
+    # NOTE! We are not keying the cache based on bundle_cache_root or
+    # fallback_roots -- the assumption here is that if you find for example
+    # <core appstore v1.2.3> this represents that particular version of some code
+    # and it doesn't matter where we are fetching it from. If <core appstore v1.2.3>
+    # is available in multiple different locations on disk, the content of each location
+    # should be identical
+
+    cache_key = str(dict_or_uri)
```

Comment:
I realize this was the previous method for generating the key, but a couple of things come to mind here.

1. Because you can represent the descriptor as either a dict or string, isn't it possible to technically cache the same thing with different keys (one dict and one uri)? Maybe that's not a big deal?

2. Can we rely on the stringified dict to be consistent regardless of how it is supplied?

If there's not too much overhead, maybe always ensure the cache key is the uri form?